### PR TITLE
serialization of arrays didn't work with non-pod types. fixed.

### PR DESCRIPTION
**Pull request by gentryx**
I did pull out some code of which the first half (related to Endianess) did just look wrong and the other half didn't work properly.

### CODE REVIEW

File: tests/unit/serialization/serialization_array.cpp [hunk 1]
```diff
@@ -199,6 +199,35 @@ void test_plain_array()
     }
 }
 
+template <typename T, std::size_t SIZE>
+void test_array_of_vectors()
```
Comment (hkaiser) [inline tests/unit/serialization/serialization_array.cpp]: Please don't use all-caps names (except single letter template parameters) - those are reserved for macros.

**Comment by hkaiser** [comment]
What's wrong with the endianess-code? if the endianess of the target does not match the one of current locality it falls back to non-optimized serialization, otherwise it invokes the optimized code path.

File: tests/unit/serialization/serialization_array.cpp (hunk 1 above)
Comment (gentryx) [inline tests/unit/serialization/serialization_array.cpp]: ACK, will fix.

**Comment by gentryx** [comment]
The Endianess-code was the same for every Endianess (or I and @sithhell are blind, which is quite possible), so that made no sense.

The fallback code invoked bitwise serialization regardless of the type of the cargo. That's not suitable for vectors (see new unit test). I only noticed this because the previous code was causing a segfault in LibGeoDecomp (because two vectors were both calling d-tors on the same memory, but the second one was doing this after the first one had already freed that memory.

**Comment by sithhell** [comment]
On 08/12/2015 06:25 PM, Hartmut Kaiser wrote:
> What's wrong with the endianess-code? if the endianess of the target
> does not match the one of current locality it falls back to
> non-optimized serialization, otherwise it invokes the optimized code path.

That's exactly what's wrong: We are in the function which is supposed to 
load/save non bitwise serializable objects, but call the functions for 
bitwise serialization if the endianness matches.

**Comment by hkaiser** [comment]
> That's exactly what's wrong: We are in the function which is supposed to  load/save non bitwise serializable objects, but call the functions for bitwise serialization if the endianness matches.

Ok, but where should this code go, then? If the endianess is not matching we shouldn't do optimized (bitwise-copy) serialization for that type.

**Comment by sithhell** [comment]
Am 12.08.2015 7:46 nachm. schrieb "Hartmut Kaiser" <notifications@github.com
>:
>>
>> That's exactly what's wrong: We are in the function which is supposed to
load/save non bitwise serializable objects, but call the functions for
bitwise serialization if the endianness matches.
>
> Ok, but where should this code go, then? If the endianess is not matching
we shouldn't do optimized (bitwise-copy) serialization for that type.

In that case, I'd say the archive has to disable array optimization
globally, that saves a lot of otherwise duplicated checks.

**Comment by hkaiser** [comment]
I think the removed code needs to be re-added to both optimized overloads, here: https://github.com/gentryx/hpx/blob/master/hpx/runtime/serialization/array.hpp#L51-L61. You were right, as it was the code was broken.

**Comment by gentryx** [comment]
I'm testing a nicer way of doing this. (Nicer than re-adding the Endianess-code twice)

**Comment by gentryx** [comment]
@hkaiser Like this?

**Comment by hkaiser** [comment]
@gentryx yes, looks good! Thanks!

**Pull request merged**